### PR TITLE
[shopsys] add frontend-api bundle with category tree

### PR DESCRIPTION
**Pull request by grossmannmartin**
| Q             | A
| ------------- | ---
|Description, reason for the PR| This PR introduces GraphQL Frontend API and use it for obtain category tree.
|New feature| Yes 
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/frontend-api/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Shopsys Frontend API
+
+[![Build Status](https://travis-ci.org/shopsys/frontend-api.svg)](https://travis-ci.org/shopsys/backend-api)
+[![Downloads](https://img.shields.io/packagist/dt/shopsys/frontend-api.svg)](https://packagist.org/packages/shopsys/frontend-api)
+
+This bundle for [Shopsys Framework](https://www.shopsys.com) adds frontend API using [overblog/GraphQLBundle](https://github.com/overblog/GraphQLBundle).
+The bundle is dedicated for projects based on Shopsys Framework (i.e. created from [`shopsys/project-base`](https://github.com/shopsys/project-base)) exclusively.
+This repository is maintained by [shopsys/shopsys] monorepo, information about changes is in [monorepo CHANGELOG.md](https://github.com/shopsys/shopsys/blob/master/CHANGELOG.md).
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/README.md]: Well, this is quite new, but I believe it should be truly master according to PR #1436. 
When version 9.0 will be released all the changes should be in the master version of the changelog file and having it differently could lead to misleading information due to viewing the not updated version.

File: packages/frontend-api/src/Resources/config/services.yml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+services:
+    _defaults:
+        autowire: true
+        autoconfigure: true
+        public: false
+
+    Shopsys\FrontendApiBundle\:
+        resource: '../../**/*{Resolver}.php'
+
+    Shopsys\FrontendApiBundle\Routing\EnabledOnCurrentDomainChecker: ~
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Resources/config/services.yml]: Good catch, thanks!

[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Resources/config/services.yml]: Well, `%shopsys_frontend_api.enabled_domains%` is not registered as parameter. 
Into this service it's passed in `ShopsysFrontendApiExtension.php`

File: upgrade/UPGRADE-v9.0.0-dev.md [hunk 3]
```diff
@@ -33,6 +33,37 @@ There you can find links to upgrade notes for other versions too.
         - `src/Resources/views/Admin/Content/Category/detail.html.twig`
         - `src/Resources/views/Admin/Content/Product/detail.html.twig`
 - add [`app/getEnvironment.php`](https://github.com/shopsys/shopsys/blob/9.0/project-base/app/getEnvironment.php) file to your project ([#1368](https://github.com/shopsys/shopsys/pull/1368))
+- introduce [Frontend API](https://github.com/shopsys/shopsys/blob/9.0/docs/frontend-api/introduction-to-frontend-api.md) to your project ([#](https://github.com/shopsys/shopsys/pull/)):
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply upgrade/UPGRADE-v9.0.0-dev.md]: Applied manually in 549f3a310

[the remark this answers is not in the document]
Comment (grossmannmartin) [reply upgrade/UPGRADE-v9.0.0-dev.md]: 👍 applied manually in 549f3a310

File: packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+     * @param int[] $enabledDomainIds
+     */
+    public function __construct(Domain $domain, array $enabledDomainIds = [])
+    {
+        $this->domain = $domain;
+        $this->enabledDomainIds = $enabledDomainIds;
+    }
+
+    /**
+     * I case no domain is selected, check if enabled on Main Admin Domain
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php]: (hopefully) improved in 156b538c1

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 5]
```diff
@@ -1,21 +1,148 @@
...
+You should see the [GraphQL documentation](https://graphql.org/learn/) for more information and learn how to query the data you need.
+
 ## Installation
-Frontend API package is installed by default in `shopsys/project-base`.
-In case you want to add the package into already existing project, you should follow [upgrade instructions](https://github.com/shopsys/shopsys/blob/9.0/upgrade/UPGRADE-v9.0.0-dev.md)
+Frontend API package is installed by default in `shopsys/project-base`.  
+In case you want to add the package into an already existing project, you should follow [upgrade instructions](https://github.com/shopsys/shopsys/blob/9.0/upgrade/UPGRADE-v9.0.0-dev.md)
 
 ## Configuration
+<!--- TODO add after configuration refactoring -->
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md]: TODO here

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+### Working with Domains
+Frontend API respects the domain you call so in case of the standard two domain setup with default data fixtures, 
+you can get data for the first domain with the request made to `http://127.0.0.1/graphql/` while requesting `http://127.0.0.2/graphql/` returns data for the second domain.
+
+### Debug your queries
+Frontend API package also integrates graphical interactive in-browser GraphQL IDE [GraphiQL](https://github.com/graphql/graphiql/tree/master/packages/graphiql#readme).
+With it, you can debug your query easily and also browse the endpoint documentation for available objects, fields and their meaning.
+
+You can access it in development mode on `http://127.0.0.1/graphql/graphiql` (respectively `http://127.0.0.1/graphql/graphiql` for the second domain).
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md]:
```suggestion
You can access it in development mode on `http://127.0.0.1/graphql/graphiql` (respectively `http://127.0.0.2/graphql/graphiql` for the second domain).
```

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+
+The `Category` object type in your project in `src/Shopsys/ShopBundle/Resources/graphql-types/Category.types.yml` is the one will be really used and you can adjust it as you want.
+`Category` is defined as
+```yaml
+Category:
+    type: object
+    inherits:
+        - 'CategoryDecorator'   # Inherits from the decorator defined earlier
+    config:
+        fields:           # Fields from the decorator are added automatically
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md]: These comments are weirdly aligned

File: packages/frontend-api/src/Resources/config/services.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/frontend-api/src/Resources/config/services.yml]: Okay that is something I do not like, because when you are looking for arguments, that are wired to the service, you look into services.yml and you find nothing, then you need to start thinking, where it might be.

File: packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+     * @param int[] $enabledDomainIds
+     */
+    public function __construct(Domain $domain, array $enabledDomainIds = [])
+    {
+        $this->domain = $domain;
+        $this->enabledDomainIds = $enabledDomainIds;
+    }
+
+    /**
+     * If no domain is selected method returns if it's enabled for the first domain instead.
```
Comment (TomasLudvik) [inline packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php]: I think it would be better to add this info somehow in methods name, because right now this behavior is not clear and you need to read that comment which you might not ...

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# Introduction to Frontend API
+
+Shopsys Framework Frontend API is an interface to the application that is dedicated for integration with your third party systems systems (like IS, storage system, ERP, ...).
```
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply docs/frontend-api/introduction-to-frontend-api.md]: resolved manually in 7fe0f9ae0

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 7 above)
Comment (grossmannmartin) [reply docs/frontend-api/introduction-to-frontend-api.md]: fixed in 34d0754fb

File: packages/frontend-api/src/Resources/config/services.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (grossmannmartin) [reply packages/frontend-api/src/Resources/config/services.yml]: I rewrote it to the standard parameter in a14a795771f082c8dbbc0d18cf4d3cc0d33eee39 and 19baa4e863e844360de88dd38dffe3d6e0d301fa
(documentation and upgrade notes are changed too in 62c5a3973b1c1d78372ed2a01da6336b1580bffe and ee698a5e1521a7dea930e907e14869dd9fa67a56)

File: docs/frontend-api/introduction-to-frontend-api.md (hunk 5 above)
Comment (grossmannmartin) [reply docs/frontend-api/introduction-to-frontend-api.md]: Content added in ee698a5e1521a7dea930e907e14869dd9fa67a56

File: packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php (hunk 8 above)
Comment (grossmannmartin) [reply packages/frontend-api/src/Routing/EnabledOnCurrentDomainChecker.php]: Changed to `isEnabledOnDomain(int $domainId)` in 4d0fda5e0b9ee10fa5af21bba41a3e2763cdf86b

File: packages/frontend-api/src/Routing/FrontendApiRouteLoader.php [hunk 10]
```diff
@@ -39,7 +48,11 @@ public function loadRoutes(): RouteCollection
     {
         $routes = new RouteCollection();
 
-        if (!$this->enabledOnCurrentDomainChecker->isEnabledOnCurrentDomain()) {
+        try {
+            if (!$this->enabledOnDomainChecker->isEnabledOnDomain($this->domain->getId())) {
+                return $routes;
+            }
+        } catch (NoDomainSelectedException $e) {
```
Comment (TomasLudvik) [inline packages/frontend-api/src/Routing/FrontendApiRouteLoader.php]: Where is this exception thrown?

Comment (grossmannmartin) [reply packages/frontend-api/src/Routing/FrontendApiRouteLoader.php]: `$this->domain->getId()` throws this exception :)

File: docs/frontend-api/introduction-to-frontend-api.md [hunk 11]
```diff
@@ -0,0 +1,157 @@
+# Introduction to Frontend API
+
+Shopsys Framework Frontend API is an interface to the application that is used for integration with external shop frontend, for example, JS storefront od mobile app.
```
Comment (TomasLudvik) [inline docs/frontend-api/introduction-to-frontend-api.md]:
```suggestion
Shopsys Framework Frontend API is an interface to the application that is used for integration with external store frontend, for example, JS storefront od mobile app.
```

**Comment by grossmannmartin** [comment]
> Hello, from testing:
> 
> * 127.0.0.1:8000/graphql does not work... It would be nice if I would not need exact address
> * if I have not enabled API on domain I will get this error: "Symfony\Component\HttpKernel\Exception\NotFoundHttpException: No route found for "GET /graphql/"" something with better description would be nice
> * all errors are returned as HTML that will be hard to parse in javascript - JSON would be highly appreciated

Thanks!

- GET request is by default redirected, but POST (request made from Postman, for example) can't be redirected and therefore the trailing slash is needed. The only way I can think of is to not specify route prefix in project-base and hardcode both variants in `FrontendApiRouteLoader`. But I'm not sure if it's worth it (especially when different behavior of routes with/without slash seems like standard behavior – http://127.0.0.1:8000/registration/ vs. http://127.0.0.1:8000/registration)
- Yes, it's standard 404 page because there is no such route. I believe it's proper behavior, because no such thing actually exist on the domain in that scenario.
- I added `ErrorHandlerListener` in ad7f2d449 to show error messages from overblog_graphql_endpoint nicely.

**Comment by grossmannmartin** [comment]
Hi @TomasLudvik,
I found a bug in enabling routes logic (caching routes could be a problem if one domain was disabled), so I changed the way how the endpoint is enabled/disabled. It's now checked at the controller level and with it, I was able to solve your comment

> if I have not enabled API on domain I will get this error: "Symfony\Component\HttpKernel\Exception\NotFoundHttpException: No route found for "GET /graphql/"" something with better description would be nice

work is done mainly in 59a7ba3 (but the last three commits are related).
Can you please check it out?

**Comment by grossmannmartin** [comment]
> * I think it is not necessary to secure access to GraphiQL in dev environment, because if you try to use it it will tell you that GraphQL is not enabled on domain. In that case I would remove services_dev.yml and other connected changes.

Thanks, you're right, it's better without the services_dev :)
Fixed in 43e1420f26ab88fa78acd148661d0e7c19ef9f59. And I also moved `EnabledOnDomainChecker` to namespace `Component` instead of `Routing`, if you agree.
